js/src/logger.ts

```diff
@@ -680,6 +682,185 @@ export interface LogOptions<IsAsyncFlush> {
+      key: `attachments/${uuid.slice(0, 2)}/${uuid}`,
+    };
+
+    this.data = this.initData(data);
+    this.uploader = this.initUploader();
+  }
+
+  /**
+   * On first access, (1) reads the attachment from disk if needed, (2)
+   * authenticates with the data plane to request a signed URL, (3) uploads to
+   * object store, and (4) updates the attachment.
+   *
+   * @returns The attachment status.
+   */
+  async upload() {
+    return await this.uploader.get();
+  }
+
+  private initUploader(): LazyValue<AttachmentStatus> {
+    const doUpload = async (conn: HTTPConnection, orgId: string) => {
+      const requestParams = new URLSearchParams({
+        key: this.reference.key,
+        filename: this.reference.filename,
+        content_type: this.reference.content_type,
+        org_id: orgId,
+      }).toString();
+      const [metadataPromiseResult, dataPromiseResult] =
+        await Promise.allSettled([
+          conn.post(`/attachment?${requestParams}`),
+          this.data.get(),
+        ]);
+      if (metadataPromiseResult.status === "rejected") {
+        const errorStr = JSON.stringify(metadataPromiseResult.reason);
+        throw new Error(
+          `Failed to request signed URL from API server: ${errorStr}`,
+        );
+      }
+      if (dataPromiseResult.status === "rejected") {
+        const errorStr = JSON.stringify(dataPromiseResult.reason);
+        throw new Error(`Failed to read file: ${errorStr}`);
+      }
+      const metadataResponse = metadataPromiseResult.value;
+      const data = dataPromiseResult.value;
+
+      let signedUrl: string | undefined;
+      try {
+        signedUrl = z
+          .object({ signedUrl: z.string().url() })
+          .parse(await metadataResponse.json()).signedUrl;
+      } catch (error) {
+        if (error instanceof ZodError) {
+          const errorStr = JSON.stringify(error.flatten());
+          throw new Error(`Invalid response from API server: ${errorStr}`);
+        }
+        throw error;
+      }
+
+      // TODO multipart upload.
+      let objectStoreResponse: Response | undefined;
+      try {
+        objectStoreResponse = await checkResponse(
+          await fetch(signedUrl, {
+            method: "PUT",
+            headers: {
+              "Content-Type": this.reference.content_type,
+            },
+            body: data,
+          }),
+        );
+      } catch (error) {
+        if (error instanceof FailedHTTPResponse) {
+          throw new Error(
+            `Failed to upload attachment to object store: ${error.status} ${error.text} ${error.data}`,
+          );
+        }
+        throw error;
+      }
+
+      return { signedUrl, metadataResponse, objectStoreResponse };
+    };
+
+    // Catches error messages and updates the attachment status.
+    const errorWrapper = async () => {
+      const status: AttachmentStatus = { upload_status: "done" };
+
+      if (!this.state) {
+        await _globalState.login({});
```

Comment:
hmm should we write it like
```
const state = this.state ?? _globalState;
await state.login({});
```

So that we log in even when they are using a separate state? It may not matter in practice if non-global states happen to always be logged in, but seems best to write it this way anyways.